File: hdl/src/Tamal/Wire.hs
-- SPDX-FileCopyrightText: 2026 Felipe Balbi
-- SPDX-License-Identifier: CERN-OHL-W-2.0

{- | The tamal wire format (design doc 2026-07-02-tamal-wire-format-design.md):
the transport-agnostic control/result byte framing. This module is the frame +
message layer over the 'Tamal.Wire.Cobs' leaf; it is a pure @[BitVector 8]@
reference model (not synthesizable — like 'Tamal.Trace.encodeRecord'). The
streaming realization is the piece-3 loader.
-}
module Tamal.Wire
  ( WireError (..)
  , ControlMsg (..)
  , wordToBytesLE
  , bytesToWordLE
  , crc8
  , frameEncode
  , frameDecode
  , encodeControl
  , decodeControl
  , encodeResult
  , decodeResult
  ) where

import Clash.Prelude

import qualified Data.List as L

import Tamal.Crc (crc8Update)
import Tamal.Wire.Cobs (cobsDecode, cobsEncode)

-- | Control-plane messages (host -> FPGA).
data ControlMsg
  = -- | instruction words to load into the instr BRAM
    LoadProgram [BitVector 32]
  | -- | Start-of-run pulse
    Trigger
  deriving stock (Generic, Show, Eq)
  deriving anyclass (NFDataX)

-- | Why a frame failed to decode.
data WireError
  = -- | CRC byte did not match the recomputed CRC
    BadCrc
  | -- | malformed COBS or missing delimiter
    BadCobs
  | -- | opcode not in the v1 set
    UnknownOpcode (BitVector 8)
  | -- | decoded frame lacks an opcode and/or CRC byte
    ShortFrame
  | -- | payload length invalid for the opcode
    BadPayloadLen
  deriving stock (Generic, Show, Eq)
  deriving anyclass (NFDataX)

{- | Split a 32-bit word into four little-endian bytes (LSB first; ISA §4).
@0xAABBCCDD -> <0xDD, 0xCC, 0xBB, 0xAA>@.
-}
wordToBytesLE :: BitVector 32 -> Vec 4 (BitVector 8)
wordToBytesLE = reverse . unpack

{- | Reassemble a little-endian byte quartet into its 32-bit word (inverse of
'wordToBytesLE').
-}
bytesToWordLE :: Vec 4 (BitVector 8) -> BitVector 32
bytesToWordLE = pack . reverse

{- | Fold the CRC-8 (poly 0x07, init 0x00, MSB-first - reuses 'Tamal.Crc') over
a whole byte sequence.
-}
crc8 :: [BitVector 8] -> BitVector 8
crc8 = L.foldl' crc8Update 0

{- | Wrap a logical frame (opcode ++ payload) for the wire: append its CRC,
COBS-encode, then append the 0x00 delimiter (§4, §8).
-}
frameEncode :: [BitVector 8] -> [BitVector 8]
frameEncode logical = cobsEncode (logical <> [crc8 logical]) <> [0]

{- | Unwrap a wire frame back to its logical bytes: strip the delimiter,
COBS-decode, split off the trailing CRC byte, and verify it.
-}
frameDecode :: [BitVector 8] -> Either WireError [BitVector 8]
frameDecode wire = do
  stripped <- stripDelim wire
  content <- maybe (Left BadCobs) Right (cobsDecode stripped)
  (logical, crc) <- splitLastByte content
  if crc8 logical == crc
    then Right logical
    else Left BadCrc

-- | Remove the trailing 0x00 delimiter (or fail).
stripDelim :: [BitVector 8] -> Either WireError [BitVector 8]
stripDelim [] = Left ShortFrame
stripDelim xs
  | L.last xs == 0 = Right (L.init xs) -- remove the last byte
  | otherwise = Left BadCobs

-- | Separate a non-empty byte list into its leading bytes and its last byte.
splitLastByte :: [BitVector 8] -> Either WireError ([BitVector 8], BitVector 8)
splitLastByte [] = Left ShortFrame
splitLastByte xs = Right (L.init xs, L.last xs)

-- | Frame opcodes (§8.1). 0x00 is the delimiter, never an opcode.
opLoadProgram :: BitVector 8
opLoadProgram = 0x01

opTrigger :: BitVector 8
opTrigger = 0x02

opTraceDrain :: BitVector 8
opTraceDrain = 0x81

-- | Encode a control message to its wire frame.
encodeControl :: ControlMsg -> [BitVector 8]
encodeControl (LoadProgram ws) =
  frameEncode (opLoadProgram : L.concatMap (toList . wordToBytesLE) ws)
encodeControl Trigger = frameEncode [opTrigger]

-- | Decode a wire frame into a control message.
decodeControl :: [BitVector 8] -> Either WireError ControlMsg
decodeControl wire = do
  logical <- frameDecode wire
  case logical of
    [] -> Left ShortFrame
    (op : payload)
      | op == opLoadProgram -> LoadProgram <$> bytesToWords payload
      | op == opTrigger ->
          if L.null payload
            then Right Trigger
            else Left BadPayloadLen
      | otherwise -> Left (UnknownOpcode op)

{- | Regroup a byte payload into little-endian 32-bit words; fail unless the
length is a multiple of four.
-}
bytesToWords :: [BitVector 8] -> Either WireError [BitVector 32]
bytesToWords [] = Right []
bytesToWords (a : b : c : d : rest) =
  (bytesToWordLE (a :> b :> c :> d :> Nil) :) <$> bytesToWords rest
bytesToWords _ = Left BadPayloadLen

{- | Encode a drained ring word-stream (REVISION ++ records ++ HALT terminator,
§8.3) as a TRACE_DRAIN wire frame.
-}
encodeResult ::
  [BitVector 32] ->
  [BitVector 8]
encodeResult ws = frameEncode (opTraceDrain : L.concatMap (toList . wordToBytesLE) ws)

-- | Decode a TRACE_DRAIN wire frame back to its ring word-stream.
decodeResult :: [BitVector 8] -> Either WireError [BitVector 32]
decodeResult wire = do
  logical <- frameDecode wire
  case logical of
    [] -> Left ShortFrame
    (op : payload)
      | op == opTraceDrain -> bytesToWords payload
      | otherwise -> Left (UnknownOpcode op)

File: crates/tamal-abi/src/wire.rs
//! Frame + message wire layer — a Rust mirror of the HDL `Tamal.Wire`.

use crate::cobs::{self, CobsError};
use crate::crc8::crc8;

/// `LOAD_PROGRAM` — host → FPGA.
pub const OP_LOAD_PROGRAM: u8 = 0x01;
/// `TRIGGER` — host → FPGA.
pub const OP_TRIGGER: u8 = 0x02;
/// `TRACE_DRAIN` — FPGA → host.
pub const OP_TRACE_DRAIN: u8 = 0x81;
/// The COBS frame delimiter.
pub const DELIMITER: u8 = 0x00;

/// A control-plane message (host → FPGA).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ControlMsg {
    /// Load N instruction words into the instruction store.
    LoadProgram(Vec<u32>),
    /// Start a run of the loaded program.
    Trigger,
}

/// Why a frame failed to decode.
#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
pub enum WireError {
    /// The trailing CRC-8 did not match.
    #[error("frame CRC mismatch")]
    BadCrc,
    /// The COBS layer was malformed.
    #[error("malformed COBS frame")]
    BadCobs,
    /// The opcode is not in the recognised set.
    #[error("unknown opcode {0:#04x}")]
    UnknownOpcode(u8),
    /// A result decode saw an opcode other than the one expected.
    #[error("wrong opcode: expected {expected:#04x}, found {found:#04x}")]
    WrongOpcode {
        /// The opcode the decoder required.
        expected: u8,
        /// The opcode actually present.
        found: u8,
    },
    /// The logical frame was shorter than `[opcode]`.
    #[error("frame too short")]
    ShortFrame,
    /// A payload / word stream was not a multiple of 4 bytes.
    #[error("payload length not a multiple of 4")]
    BadPayloadLen,
}

impl From<CobsError> for WireError {
    fn from(_: CobsError) -> Self {
        WireError::BadCobs
    }
}

/// Pack words little-endian: `0xAABBCCDD -> [DD, CC, BB, AA]`.
pub fn words_to_le_bytes(words: &[u32]) -> Vec<u8> {
    let mut out = Vec::with_capacity(words.len() * 4);
    for w in words {
        out.extend_from_slice(&w.to_le_bytes());
    }
    out
}

/// Unpack little-endian words; `BadPayloadLen` if `bytes.len() % 4 != 0`.
pub fn le_bytes_to_words(bytes: &[u8]) -> Result<Vec<u32>, WireError> {
    if bytes.len() % 4 != 0 {
        return Err(WireError::BadPayloadLen);
    }
    Ok(bytes
        .chunks_exact(4)
        .map(|c| u32::from_le_bytes([c[0], c[1], c[2], c[3]]))
        .collect())
}

/// Wrap a logical frame: `COBS(logical ++ crc8(logical)) ++ 0x00`.
pub fn frame_encode(logical: &[u8]) -> Vec<u8> {
    let mut with_crc = logical.to_vec();
    with_crc.push(crc8(logical));
    let mut wire = cobs::cobs_encode(&with_crc);
    wire.push(DELIMITER);
    wire
}

/// Unwrap a wire frame to its logical bytes (`opcode ++ payload`), verifying CRC.
pub fn frame_decode(wire: &[u8]) -> Result<Vec<u8>, WireError> {
    let body = match wire.split_last() {
        Some((&DELIMITER, body)) => body,
        _ => return Err(WireError::BadCobs), // missing delimiter
    };
    let decoded = cobs::cobs_decode(body)?;
    let (crc, logical) = decoded.split_last().ok_or(WireError::ShortFrame)?;
    if logical.is_empty() {
        return Err(WireError::ShortFrame);
    }
    if crc8(logical) != *crc {
        return Err(WireError::BadCrc);
    }
    Ok(logical.to_vec())
}

/// Encode a control message to wire bytes.
pub fn encode_control(msg: &ControlMsg) -> Vec<u8> {
    let logical = match msg {
        ControlMsg::Trigger => vec![OP_TRIGGER],
        ControlMsg::LoadProgram(words) => {
            let mut l = Vec::with_capacity(1 + words.len() * 4);
            l.push(OP_LOAD_PROGRAM);
            l.extend_from_slice(&words_to_le_bytes(words));
            l
        }
    };
    frame_encode(&logical)
}

/// Decode a control frame (used by tests / a future device-side simulator).
pub fn decode_control(wire: &[u8]) -> Result<ControlMsg, WireError> {
    let logical = frame_decode(wire)?;
    match logical[0] {
        OP_LOAD_PROGRAM => Ok(ControlMsg::LoadProgram(le_bytes_to_words(&logical[1..])?)),
        OP_TRIGGER => {
            if logical.len() != 1 {
                return Err(WireError::BadPayloadLen);
            }
            Ok(ControlMsg::Trigger)
        }
        other => Err(WireError::UnknownOpcode(other)),
    }
}

/// Decode a `TRACE_DRAIN` result frame into little-endian words.
pub fn decode_result(wire: &[u8]) -> Result<Vec<u32>, WireError> {
    let logical = frame_decode(wire)?;
    if logical[0] != OP_TRACE_DRAIN {
        return Err(WireError::WrongOpcode {
            expected: OP_TRACE_DRAIN,
            found: logical[0],
        });
    }
    le_bytes_to_words(&logical[1..])
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;

    #[test]
    fn le_round_trip_vector() {
        assert_eq!(
            words_to_le_bytes(&[0xAABB_CCDD]),
            vec![0xDD, 0xCC, 0xBB, 0xAA]
        );
        assert_eq!(
            le_bytes_to_words(&[0xDD, 0xCC, 0xBB, 0xAA]).unwrap(),
            vec![0xAABB_CCDD]
        );
    }

    #[test]
    fn le_bytes_len_must_be_multiple_of_four() {
        assert_eq!(
            le_bytes_to_words(&[0x01, 0x02, 0x03]),
            Err(WireError::BadPayloadLen)
        );
    }

    #[test]
    fn trigger_encodes_to_pinned_bytes() {
        // logical [0x02, crc8([0x02])=0x0E] -> COBS [0x03,0x02,0x0E] -> +delim.
        assert_eq!(
            encode_control(&ControlMsg::Trigger),
            vec![0x03, 0x02, 0x0E, 0x00]
        );
    }

    #[test]
    fn load_program_round_trips_through_frame_decode() {
        let words = vec![0x0800_0064u32, 0x0000_0000u32];
        let wire = encode_control(&ControlMsg::LoadProgram(words.clone()));
        assert_eq!(*wire.last().unwrap(), 0x00, "ends in delimiter");
        assert_eq!(
            wire[..wire.len() - 1].iter().filter(|&&b| b == 0).count(),
            0,
            "no interior zero"
        );
        let logical = frame_decode(&wire).unwrap();
        assert_eq!(logical[0], OP_LOAD_PROGRAM);
        assert_eq!(le_bytes_to_words(&logical[1..]).unwrap(), words);
    }

    #[test]
    fn decode_result_unpacks_drain_words() {
        let words = vec![0x0001_0000u32, 0xC000_0000u32];
        let mut logical = vec![OP_TRACE_DRAIN];
        logical.extend_from_slice(&words_to_le_bytes(&words));
        let wire = frame_encode(&logical);
        assert_eq!(decode_result(&wire).unwrap(), words);
    }

    #[test]
    fn single_byte_flip_fails_decode() {
        let wire = encode_control(&ControlMsg::LoadProgram(vec![0x1234_5678]));
        for i in 0..wire.len() - 1 {
            let mut bad = wire.clone();
            bad[i] ^= 0x01;
            // Either the delimiter scan, COBS, or CRC rejects it — never a silent wrong decode.
            let _ = frame_decode(&bad); // must not panic
        }
        // A flipped payload byte under a good frame shape must trip the CRC.
        let mut bad = wire.clone();
        bad[2] ^= 0x01;
        assert!(matches!(
            frame_decode(&bad),
            Err(WireError::BadCrc) | Err(WireError::BadCobs)
        ));
    }

    #[test]
    fn error_taxonomy() {
        // Unknown opcode.
        let logical = vec![0x7Fu8];
        let wire = frame_encode(&logical);
        assert_eq!(decode_control(&wire), Err(WireError::UnknownOpcode(0x7F)));
        // decode_result on a non-0x81 frame.
        let wire = encode_control(&ControlMsg::Trigger);
        assert_eq!(
            decode_result(&wire),
            Err(WireError::WrongOpcode {
                expected: OP_TRACE_DRAIN,
                found: OP_TRIGGER
            })
        );
        // Short frame: a bare delimiter decodes to an empty logical frame.
        assert_eq!(frame_decode(&[0x01, 0x00]), Err(WireError::ShortFrame));
    }

    proptest! {
        #[test]
        fn load_program_word_round_trip(ws in prop::collection::vec(any::<u32>(), 0..40)) {
            let wire = encode_control(&ControlMsg::LoadProgram(ws.clone()));
            let logical = frame_decode(&wire).unwrap();
            prop_assert_eq!(logical[0], OP_LOAD_PROGRAM);
            prop_assert_eq!(le_bytes_to_words(&logical[1..]).unwrap(), ws);
        }
    }
}
